### PR TITLE
Docs: Add IIS documentation

### LINKED ISSUE #649: Add examples on how to configure servers in order to pass the rules
can you show me an example for a .htaccess which passes your scanner rules?
this doesnt pass your rules

```apache
 Header set X-UA-Compatible "IE=Edge,chrome=1"
 Header set X-Frame-Options "SAMEORIGIN"
 Header set X-XSS-Protection "1; mode=block"
 Header set X-Content-Type-Options: "nosniff"
 Header set Strict-Transport-Security "max-age=31536000; includeSubDomains;"
 Header set Referrer-Policy "strict-origin-when-cross-origin"
```

### PR DESCRIPTION
**Pull request by molant**
## Pull request checklist

Make sure you:

- [X] Signed the [Contributor License Agreement](https://cla.js.foundation/sonarwhal/sonarwhal)
- [X] Followed the [commit message guidelines](https://sonarwhal.com/docs/contributor-guide/contributing/pull-requests/#commitmessages)

For non-trivial changes, please make sure you also:

- [X] Added/Updated related documentation.

## Short description of the change(s)

Add examples on how to configure IIS to pass certain rules.

Fix #649

### CODE REVIEW

File: docs/user-guide/rules/content-type.md [hunk 1]
```diff
@@ -286,6 +286,81 @@ Note that:
 
 </details>
 
+<details>
+<summary>How to configure IIS</summary>
+
+By default IIS [maps certain filename extensions to specific media
+types][mime.types iis(https://support.microsoft.com/en-us/help/936496/description-of-the-default-settings-for-the-mimemap-property-and-for-t)], but depending on the IIS version that is
```
Comment (alrra) [inline docs/user-guide/rules/content-type.md]: Fix link.

File: docs/user-guide/rules/content-type.md [hunk 2]
```diff
@@ -286,6 +286,81 @@ Note that:
...
+used, some mappings may be outdated or missing.
+
+Fortunately, IIS provides a way to overwrite and add to the existing
+media types mappings using the [`<mimeMap>` element under <staticContent>][mimeMap(https://docs.microsoft.com/en-us/iis/configuration/system.webserver/staticcontent/mimemap)]. For
```
Comment (alrra) [inline docs/user-guide/rules/content-type.md]: Fix link.

File: docs/user-guide/rules/highest-available-document-mode.md [hunk 3]
```diff
@@ -283,6 +283,77 @@ Note that:
 
 </details>
 
+<details>
+<summary>How to configure IIS</summary>
+
+### Adding the `X-UA-Compatible` header
+
+To add the `X-UA-Compatible` header you can use a [`URL rewrite` rule][url rewrite]
+that matches the `text/html` content-type of a response and then adds it:
```
Comment (alrra) [inline docs/user-guide/rules/highest-available-document-mode.md]: `application/xhtml+xml` is also commonly used.

Comment (alrra) [reply docs/user-guide/rules/highest-available-document-mode.md]: content-type => media type

File: docs/user-guide/rules/highest-available-document-mode.md [hunk 4]
```diff
@@ -283,6 +283,77 @@ Note that:
...
+```xml
+<configuration>
+     <system.webServer>
+        <httpProtocol>
+             <customHeaders>
+                <add name="X-UA-Compatible" value="IE=edge"/>
+             </customHeaders>
+         </httpProtocol>
+    </system.webServer>
+</configuration>
```
Comment (alrra) [inline docs/user-guide/rules/highest-available-document-mode.md]: I don't think we should include this example.

File: docs/user-guide/rules/http-cache.md [hunk 5]
```diff
@@ -279,6 +279,52 @@ Also note that:
 
 </details>
 
+<details>
+<summary>How to configure IIS</summary>
+
+You can enable the `Cache-Control` and/or `Expire` headers on IIS using
+the [`<clientCache> element under <staticContent>`][clientcache iis(https://docs.microsoft.com/en-us/iis/configuration/system.webserver/staticcontent/clientcache)].
```
Comment (alrra) [inline docs/user-guide/rules/http-cache.md]: Fix link.

File: docs/user-guide/rules/http-cache.md [hunk 6]
```diff
@@ -279,6 +279,52 @@ Also note that:
...
+            <staticContent>
+                <clientCache cacheControlMode="UseMaxAge" cacheControlMaxAge="365.00:00:00" cacheControlCustom="immutable" />
+            </staticContent>
+        </system.webServer>
+    </location>
+</configuration>
+```
+
+In this case, you would want to have your JavaScript, CSS, images, etc.
+under `/static`, and your HTML elesewhere.
```
Comment (alrra) [inline docs/user-guide/rules/http-cache.md]: Maybe say something in the lines of: If your content is not under `static/`, you will need to change the `<location path="static">` part and point it to the right directory?

File: docs/user-guide/rules/http-cache.md [hunk 7]
```diff
@@ -279,6 +279,52 @@ Also note that:
...
+
+Important notes:
+
+* Do not use the above snippet if you are not doing filename revving.
+* The above snippet works with IIS 6+.
+* You should use the above snippet in the [`web.config` of your
+  application][iis conf file(FIND LINK)] under
```
Comment (alrra) [inline docs/user-guide/rules/http-cache.md]: Find link.

File: docs/user-guide/rules/http-compression.md [hunk 8]
```diff
@@ -795,6 +795,164 @@ Also note that:
 
 </details>
 
+<details>
+<summary>How to configure IIS</summary>
+
+IIS can be configured to conditionally (based on media type)
+compress resources using gzip as well as send the appropriate
+`Content-Encoding` and `Vary` headers using [`mod_deflate`][mod_deflate]
```
Comment (alrra) [inline docs/user-guide/rules/http-compression.md]: This needs to be updated.

File: docs/user-guide/rules/no-disallowed-headers.md [hunk 9]
```diff
@@ -143,6 +143,65 @@ Note that:
...
+
+The following snippet will remove the `X-Powered-By` headers from all
+responses:
+
+```xml
+<configuration>
+     <system.webServer>
+        <httpProtocol>
+             <customHeaders>
+                <remove name="X-Powered-By"/>
```
Comment (alrra) [inline docs/user-guide/rules/no-disallowed-headers.md]: Maybe include all the headers that we disallow by default?

File: docs/user-guide/rules/highest-available-document-mode.md [hunk 10]
```diff
@@ -226,7 +226,7 @@ X-UA-Compatible: ie=edge
 Apache can be configured to add or remove the `X-UA-Compatible`
 header using the [`Header` directive][header directive].
 
-### Adding the `X-UA-Compatible` header
+### Adding the `X-UA-Compatible` header on Apache
```
Comment (molant) [inline docs/user-guide/rules/highest-available-document-mode.md]: Otherwise the linter complains that we have 2 titles with the same name

**Comment by molant** [comment]
OK, this is ready for review!

File: docs/user-guide/rules/no-disallowed-headers.md [hunk 11]
```diff
@@ -143,6 +143,65 @@ Note that:
...
+                    <match serverVariable="RESPONSE_Server" pattern=".+" />
+                    <action type="Rewrite" value="" />
+                </rule>
+            </outboundRules>
+        </rewrite>
+    </system.webServer>
+</configuration>
+```
+
+The above snippet will use a [`URL rewrite`][url rewrite] rule to remove
```
Comment (alrra) [inline docs/user-guide/rules/no-disallowed-headers.md]: Does IIS have this module enabled by default or users usually need to enable it?

If it's always enabled by default, then I don't think we should specify this.

File: docs/user-guide/rules/no-html-only-headers.md [hunk 12]
```diff
@@ -130,6 +130,56 @@ Note that:
...
+If you are not passing the rule because a non HTML resource has them
+you are probably doing something like the following that adds it
+unconditionally:
+
+```xml
+<configuration>
+     <system.webServer>
+        <httpProtocol>
+             <customHeaders>
+                <add name="X-UA-Compatible" value="IE=edge"/>
```
Comment (alrra) [inline docs/user-guide/rules/no-html-only-headers.md]: Again I think we should include all the headers that we disallow by default.

File: docs/user-guide/rules/no-html-only-headers.md [hunk 13]
```diff
@@ -130,6 +130,56 @@ Note that:
...
+
+```xml
+<configuration>
+     <system.webServer>
+        <rewrite>
+            <outboundRules>
+                <rule name="X-UA-Compatible header">
+                    <match serverVariable="RESPONSE_X_UA_Compatible" pattern=".*" />
+                    <conditions>
+                        <add input="{RESPONSE_CONTENT_TYPE}" pattern="^text/html" />
```
Comment (alrra) [inline docs/user-guide/rules/no-html-only-headers.md]: Again, I think we should also add `application/xhtml+xml`.

File: docs/user-guide/rules/highest-available-document-mode.md (hunk 3 above)
Comment (molant) [reply docs/user-guide/rules/highest-available-document-mode.md]: > content-type => media type

I'm matching the `content-type` header.

Comment (molant) [reply docs/user-guide/rules/highest-available-document-mode.md]: > `application/xhtml+xml` is also commonly used.

I'll make a note bellow replacing. Adding rules and conditions has a perf impact so user should try to make the rewrite rules as lean as possible.

File: docs/user-guide/rules/highest-available-document-mode.md (hunk 4 above)
Comment (molant) [reply docs/user-guide/rules/highest-available-document-mode.md]: OK

File: docs/user-guide/rules/http-cache.md (hunk 7 above)
Comment (molant) [reply docs/user-guide/rules/http-cache.md]: It was a leftover

**Comment by molant** [comment]
Ready for another round

File: docs/user-guide/rules/highest-available-document-mode.md [hunk 14]
```diff
@@ -283,6 +283,63 @@ Note that:
...
+                    </conditions>
+                    <action type="Rewrite" value="IE=edge"/>
+                </rule>
+            </outboundRules>
+        </rewrite>
+    </system.webServer>
+</configuration>
+```
+
+Note that if your site uses a mime type different than `text/html` to serve HTML
```
Comment (alrra) [inline docs/user-guide/rules/highest-available-document-mode.md]: > mime type different than `text/html`  (e.g.: `application/xhtml+xml`)....

File: docs/user-guide/rules/highest-available-document-mode.md [hunk 15]
```diff
@@ -283,6 +283,63 @@ Note that:
...
+content you'll have to update the value of `pattern` in
+`<add input="{RESPONSE_CONTENT_TYPE}" pattern="^text/html" />`.
```
Comment (alrra) [inline docs/user-guide/rules/highest-available-document-mode.md]: I think this line can be dropped.

File: docs/user-guide/rules/content-type.md [hunk 16]
```diff
@@ -286,6 +286,80 @@ Note that:
...
+            <mimeMap fileExtension=".txt" mimeType="text/plain; charset=utf-8" />
+            <remove fileExtension=".xml"/>
+            <mimeMap fileExtension=".xml" mimeType="text/xml; charset=utf-8"/>
+            <remove fileExtension=".webmanifest"/>
+            <mimeMap fileExtension="webmanifest" mimeType="application/manifest+json; charset=utf-8"/>
+            <!-- font types -->
+            <remove fileExtension=".woff"/>
+            <mimeMap fileExtension=".woff" mimeType="font/woff"/>
+            <remove fileExtension=".woff2"/>
+            <mimeMap fileExtension=".woff2" mimeType="font/woff2"/>
```
[the remark this answers is not in the document]
Comment (molant) [reply docs/user-guide/rules/content-type.md]: Done

File: docs/user-guide/rules/http-compression.md [hunk 17]
```diff
@@ -795,6 +795,162 @@ Also note that:
...
+                    <add key="webmanifest.gz" value="application/manifest+json; charset=utf-8" />
+                    <add key="css.br" value="text/css; charset=utf-8" />
+                    <add key="html.br" value="text/html; charset=utf-8" />
+                    <add key="ico.br" value="image/x-icon" />
+                    <add key="js.br" value="text/javascript; charset=utf-8" />
+                    <add key="map.br" value="application/json; charset=utf-8" />
+                    <add key="svg.br" value="image/svg+xml; charset=utf-8" />
+                    <add key="txt.br" value="text/plain; charset=utf-8" />
+                    <add key="xml.br" value="text/xml; charset=utf-8" />
+                    <add key="webmanifest.br" value="application/manifest+json; charset=utf-8" />
```
Comment (alrra) [inline docs/user-guide/rules/http-compression.md]: You are missing a few ones (same as with the media types).

File: docs/user-guide/rules/http-compression.md [hunk 18]
```diff
@@ -795,6 +790,183 @@ Also note that:
...
+  * the original (e.g.: script.js) - you should also have this file
+    in case the user agent doesn't requests things compressed
+  * the file compressed with Zopfli (e.g.: script.js.gz)
+  * the file compressed with Brotli (e.g.: script.js.br)
+
+```xml
+<configuration>
+    <system.webServer>
+        <staticContent>
+            <!-- IIS doesn't know about brotli. If mimeMap is not added
```
Comment (alrra) [inline docs/user-guide/rules/http-compression.md]: **B**rotli

File: docs/user-guide/rules/x-content-type-options.md [hunk 19]
```diff
@@ -117,6 +117,34 @@ Note that:
...
+Presuming the script files are sent with the `Content-Type` header set
+to `text/javascript` and styleshees to `text/css` you can use
+a [`URL rewrite` rule][url rewrite] like the following:
+
+```xml
+<configuration>
+     <system.webServer>
+        <rewrite>
+            <outboundRules>
+                <!-- Add X-Content-Type to text/javascript and text/css responses -->
```
Comment (alrra) [inline docs/user-guide/rules/x-content-type-options.md]: X-Content-Type-Options

**Pull request closed without merging**